### ISSUE TITLE
Should we enable F16C instructions in CUTLASS?

### ISSUE THREAD

**Issue by Skylion007** [pytorch/pytorch#132579]
### 🚀 The feature, motivation and pitch

I was exploring options in the CUTLASS CMakelist as we may need to turn on https://github.com/NVIDIA/cutlass/blob/06b21349bcf6ddf6a1686a47a137ad1446579db9/CMakeLists.txt#L313 for flash attention 3 and general performance boosts and found CUTLASS_ENABLE_F16C https://github.com/NVIDIA/cutlass/blob/06b21349bcf6ddf6a1686a47a137ad1446579db9/CMakeLists.txt#L313 which is defaulted to off. Now I couldn't tell whether this instruction set is supported by PyTorch as a minimum requirement at the moment, but I believe it is for X86 architecture. If we do require this instruction set anyway for CPU ops, we should consider enabling this flag.

### Alternatives

* Leave it turned off and eat the minor performance loss for better compatibility with really old X86 archs.
* Turn it on conditional for the inductor CUTLASS backend if the CPU on the host supports it.

**Comment by cyberluke** [comment]
Of course it should be enabled!

F16C is supported since Ivy Bridge - 2012; 12 years ago

This will speed up conversion between float16 and float32.

Nobody use x86, you cannot even run Ubuntu on x86 anymore.

**Comment by cyberluke** [comment]
But I don't like this code in CUTLASS CMakeList:
```
if (CUTLASS_ENABLE_F16C AND NOT CMAKE_CROSSCOMPILING)
  list(APPEND CUTLASS_CUDA_FLAGS -DCUTLASS_ENABLE_F16C=1)
  if ((CMAKE_CXX_COMPILER_ID MATCHES "GNU") OR (CMAKE_CXX_COMPILER_ID MATCHES "Clang"))
    list(APPEND CUTLASS_CUDA_NVCC_FLAGS -Xcompiler=-mf16c)
  elseif((CMAKE_CXX_COMPILER_ID MATCHES "MSVC"))
    list(APPEND CUTLASS_CUDA_NVCC_FLAGS -Xcompiler=/arch:AVX2)
  endif()
endif()
```

For GNU compiler (Linux/Unix), it will correctly use F16C compiler parameter, but for MSVC (Microsoft Windows), it will force AVX2 architecture, which is not correct. 

I have i7-4930K, I have F16C support with AVX, but not AVX2. In that case I could not run it, but I could use WSL (Windows Linux Subsystem) and run Linux Docker container with F16C on Windows.

Conclusion:
So the best option is to enable F16C, but don't use MSVC compiler. Use GNU compiler for all platforms. For Windows I recommend compiling with Mingw-w64 https://www.mingw-w64.org/

**Comment by Skylion007** [comment]
i don't think we ever actually use the CUTLASS CMAKE, we directly include the headers, which honestly may be getting worse and worse as the options for CUTLASS become more and more complicated.
